apiservice/apiservice.go

```diff
@@ -0,0 +1,268 @@
+package apiservice
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow/protobuf/go/flow/access"
+
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/grpcutils"
+)
+
+func NewFlowAPIService(accessNodeAddressAndPort flow.IdentityList, timeout time.Duration) (*FlowAPIService, error) {
+	accessClients := make([]access.AccessAPIClient, accessNodeAddressAndPort.Count())
+	for i, identity := range accessNodeAddressAndPort {
+		if identity.NetworkPubKey == nil {
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithInsecure(), //nolint:staticcheck
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return nil, err
+			}
+
+			accessClients[i] = access.NewAccessAPIClient(clientRPCConnection)
+		} else {
+			tlsConfig, err := grpcutils.DefaultClientTLSConfig(identity.NetworkPubKey)
+			if err != nil {
+				return nil, err
+			}
+
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return nil, err
+			}
+
+			accessClients[i] = access.NewAccessAPIClient(clientRPCConnection)
+		}
+	}
+
+	ret := &FlowAPIService{}
+	ret.upstream = accessClients
+	ret.LocalCache = nil
+	ret.roundRobin = 0
+	return ret, nil
+}
+
+type FlowAPIService struct {
+	access.UnimplementedAccessAPIServer
+	roundRobin int
+	upstream   []access.AccessAPIClient
+	LocalCache access.AccessAPIServer
+}
+
+func (h *FlowAPIService) Ping(context context.Context, req *access.PingRequest) (*access.PingResponse, error) {
+	if h.LocalCache != nil {
```

Comment:
re: my previous comment about `LocalCache` always being set, I think we should assume that `LocalCache` is configured. If there's a bug, the node will crash with an obvious panic pointing to the issue, which should be easy to catch with testing.

Reply:
I fixed this in a previous comment.
I remember that you objected the LocalCache null check.
All null checks are already removed from the codebase.